src/main/java/com/minecolonies/coremod/tileentities/TileEntityRack.java

```diff
@@ -0,0 +1,473 @@
+        }
+    };
+
+    /**
+     * Check if a certain itemstack is present in the inventory.
+     * This method checks the content list, it is therefore extremely fast.
+     *
+     * @param stack the stack to check.
+     * @return true if so.
+     */
+    public boolean hasItemStack(final ItemStack stack)
+    {
+        return content.containsKey(new ItemStorage(stack));
+    }
+
+    /**
+     * Checks if the chest is empty.
+     * This method checks the content list, it is therefore extremely fast.
+     *
+     * @return true if so.
+     */
+    public boolean isEmpty()
+    {
+        return content.isEmpty();
+    }
+
+    /**
+     * Checks if the chest is empty.
+     * This method checks the content list, it is therefore extremely fast.
+     *
+     * @return true if so.
+     */
+    public boolean freeStacks()
+    {
+        return content.isEmpty();
+    }
+
+    /**
+     * Get the amount of free slots in the inventory.
+     * This method checks the content list, it is therefore extremely fast.
+     *
+     * @return the amount of free slots (an integer).
+     */
+    public int getFreeSlots()
+    {
+        int freeSlots = inventory.getSlots();
+        for (int itemAmount : content.values())
+        {
+            final double slotsNeeded = (double) itemAmount / Constants.STACKSIZE;
+            freeSlots -= (int) Math.ceil(slotsNeeded);
+        }
+        return freeSlots;
+    }
+
+    /**
+     * Check if a similar/same item as the stack is in the inventory.
+     * This method checks the content list, it is therefore extremely fast.
+     *
+     * @param stack             the stack to check.
+     * @param ignoreDamageValue ignore the damage value.
+     * @return true if so.
+     */
+    public boolean hasItemStack(final ItemStack stack, final boolean ignoreDamageValue)
+    {
+        final ItemStorage compareStorage = new ItemStorage(stack, ignoreDamageValue);
+        for (final Map.Entry<ItemStorage, Integer> entry : content.entrySet())
+        {
+            if (compareStorage.equals(entry.getKey()))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Check if a similar/same item as the stack is in the inventory.
+     * This method checks the content list, it is therefore extremely fast.
+     *
+     * @param itemStackSelectionPredicate the predicate to test the stack against.
+     * @return true if so.
+     */
+    public boolean hasItemStack(@NotNull final Predicate<ItemStack> itemStackSelectionPredicate)
+    {
+        for (final Map.Entry<ItemStorage, Integer> entry : content.entrySet())
+        {
+            if (itemStackSelectionPredicate.test(entry.getKey().getItemStack()))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Scans through the whole storage and updates it.
+     */
+    public void updateItemStorage()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "updateItemStorage" is 16 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 26 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)